### PR TITLE
[ember-data] Fix Model.belongsTo and Model.hasMany

### PR DESCRIPTION
**Pull request by ro0gr**
This enables relationships aware `model.belongsTo(` and `model.hasMany(`:
 - Specifying correct relationship name correctly infers related model type via `reference.value()`
 - fixed `this.belongsTo` and `this.hasMany`, which have been broken previously
 - Specifying wrong key for relationship now have one of the following behaviors:
    - passing a key which doesn't exist on a model, would lead to TS error
    - passing a key which is defined, but doesn't match relationship kind(for instance `this.belongsTo` for `hasMany` field), would result to `DS.BelongsToReference<never>` or `DS.HasManyReference<never>`, which is not a hard error, but prevents you to use any of reference apis. I've tried to achieve a hard error here, but it doesn't seem super easy to do, so I left it in a state described above.

----

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
   - [WIP docs for native classes](https://github.com/typed-ember/ember-cli-typescript/blob/657f25cad01e1f725152257cd625d20e0606c175/tests/dummy/app/templates/docs/ember-data/models.md), PR - https://github.com/typed-ember/ember-cli-typescript/pull/935
   - https://api.emberjs.com/ember-data/3.16/functions/@ember-data%2Fmodel/belongsTo
   - https://api.emberjs.com/ember-data/release/classes/BelongsToReference
   - https://api.emberjs.com/ember-data/3.16/functions/@ember-data%2Fmodel/hasMany
   - https://api.emberjs.com/ember-data/release/classes/HasManyReference

### CODE REVIEW

File: types/ember-data/index.d.ts [hunk 1]
```diff
@@ -23,7 +23,30 @@ import AdapterRegistry from 'ember-data/types/registries/adapter';
...
-type RelationshipsFor<Model extends DS.Model> = ModelKeys<Model>; // TODO: filter to hasMany/belongsTo properties only (TS 2.8)
+type RelationshipKeysFor<Model extends DS.Model> = keyof RelationshipsFor<Model>;
+
+type RelationshipsFor<Model extends DS.Model> = HasManyRelationshipsFor<Model> & BelongsToRelatipsFor<Model>;
+
+type BelongsToRelatipsFor<Model extends DS.Model> = {
+    [k in keyof Model]:
+      Model[k] extends Ember.ComputedProperty<infer C & DS.PromiseObject<infer C>, infer C>
+        ? C extends DS.Model ? C : never
+      : Model[k] extends DS.Model ? Model[k]
```
Comment (ro0gr) [inline types/ember-data/index.d.ts]: this line stands to handle decorators version. Should probably add 
```ts
: Model[k] extends DS.PromiseObject<infer C> ? C
```
to support 
```ts
export default class Post extends Model {
  @belongsTo('user')
  user!: DS.PromiseObject<User>;
}
```

File: types/ember-data/index.d.ts [hunk 2]
```diff
@@ -23,7 +23,30 @@ import AdapterRegistry from 'ember-data/types/registries/adapter';
...
+      : never
+  };
+
+type HasManyRelationshipsFor<Model extends DS.Model> = {
+    [k in keyof Model]:
+      Model[k] extends Ember.ComputedProperty<DS.PromiseManyArray<infer C>, Ember.Array<infer C>>
+        ? C extends DS.Model ? C : never
+      : Model[k] extends Ember.ComputedProperty<DS.ManyArray<infer C>>
+        ? C extends DS.Model ? C : never
+      : Model[k] extends DS.PromiseManyArray<infer C>
```
Comment (ro0gr) [inline types/ember-data/index.d.ts]: should probably add 
```ts
Ember.Array<infer C> ? C extends DS.Model ? C 
  : never`

File: types/ember-data__model/tsconfig.json [hunk 3]
```diff
@@ -1,6 +1,7 @@
 {
     "compilerOptions": {
         "module": "commonjs",
+        "target": "ES5",
```
Comment (ro0gr) [inline types/ember-data__model/tsconfig.json]: ✂️ not needed anymore

**Comment by ro0gr** [comment]
ok, I've detected some issues while adding more coverage. Marking as a WIP for now, will to figure it out soon.

**Comment by ro0gr** [comment]
Just checked it against my current mid-size code base. There are few breakages detected. All are related to methods defined on models, which use `this.belongsTo(` and and `this.hasMany`. For the record, these are pre-existing issues, which just started to appear in new ways.

1. 
```ts
this.belongsTo('rel' as any).value()?.id;
```

Error:
> Property 'id' does not exist on type 'NonNullable<UnpackBelongsTo<this>[any]>'

*Solution*: remove explicit `this`. Not sure if that's a big issue TBH.

2.

```ts
  @belongsTo('something')
  something!: Something

  get explicitThis(this: Me) {
    return this.belongsTo('something').value();
  }
```

Error:

> 'explicitThis' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.

Solution: remove explicit `this`. It doesn't seem a big issue as well. However, at the moment, I'm not sure what is the root cause. To me it should just work, even if explicit `this` is not a cool thing.

3.

```ts
class Parent extends Model {
  @hasMany('child')
  children!: Child

  get firstFromHasMany() {
    const p = this.hasMany('children').value();

    return p && get(p, 'firstObject');
  }
}
```

Error:
> 'firstFromHasMany' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.ts(7022)

Solution: declare explicit `this` and the return type. While this is better than `as any` imo, I'd like to understand if we can do smth with it before(if) merging the PR. 

```ts
class Parent extends Model {
  @hasMany('child')
  children!: Child

  get firstFromHasMany(this: Parent): Child | undefined {
    const p = this.hasMany('children').value();

    return p && get(p, 'firstObject');
  }
}
```  
----

The good news are that everything else seem to work as expected at the moment, which even helped me to find some suspicious places in my codebase :)

**Comment by chriskrycho** [comment]
Thanks for this! I'm going to try to spend some time on it today. cc. also @dfreeman @jamescdavis who may be able to get eyes on it.

**Comment by ro0gr** [comment]
> I'd like to understand if we can do smth with it before(if) merging the PR.

After some brute-forcing, I think what we have with a broken `this.hasMany` is TS bug. Basically, in order to make work `hasMany`, we do smth like(extremelly reduced version):

```ts
type UnpackHasMany <T> = {
    [k in keyof T]: T[k] extends Array<infer U>
      ? U
      : T
};

declare class Model {
    hasMany<T extends Model, K extends keyof UnpackHasMany<T>>(this: T, name: K): UnpackHasMany<T>[K];
}

class A extends Model {
    myHasMany!: Array<{ name: string }>; // array stands to mimic hasMany

    test() {
        // Fails with:
        // Property 'name' does not exist on type 'UnpackHasMany<this>["myHasMany"]'.
        return this.hasMany('myHasMany').name;
    }
}

const a = new A();

// meanwhile it works as expected for instances.
a.hasMany('myHasMany'); // $ExpectType number

```

But, if I extract infering an array item type to a separate type, which deals with `T` instead of `T[k]` it starts working:

```ts
// this one solves the issue.
type UnboxedHasMany<T> = T extends Array<infer U> ? U : T;

type UnpackHasMany <T> = {
    [k in keyof T]: UnboxedHasMany<T[k]>
};

class A extends Model {
    myHasMany!: Array<{ name: string }>; // array stands to mimic hasMany

    test(): string {
        // `this` works as expected!
        return this.hasMany('myHasMany').name;
    }
}
```

I'm not 100% sure about root causes of such behavior, but I'll update PR to fix `this.hasMany` and cover it by tests.

**Comment by ro0gr** [comment]
I believe I've addressed my concerns. Please review, and let me know if something else should have been done here, when you have time.

**Comment by ro0gr** [comment]
alright, I've found some issues with `assertType` which is demonstrated and fixed in the last 3 commits.
So first, I've broken relationships unpacking, but test like:
```ts
  assertType<Folder|null>(
    folder.belongsTo('parent').value() // because of intentionally broken styles it's "never"
  );
```
remained green.

That made me to re-write some of the tests to `$ExpectType` flavour.

However... Now, even with CI passing, and my editor doesn't highlight any issues, `npm run lint ember-data` fails for TypeScript@next, with errors like:

```ts
Error: ~/workspace/DefinitelyTyped/types/ember-data/test/belongs-to.ts:32:1
ERROR: 32:1  expect  TypeScript@next expected type to be:
  Folder
got:
  Model & Folder
```

And I'm pretty clueless at the moment. I'd like someone to check the PR locally with `npm run lint ember-data`, and let me know if they see the same issue. Would appreciate any hints.

**Comment by chriskrycho** [comment]
Thanks so much for pushing so thoroughly on this! I will do my best to give it some love and time this coming week. Perhaps @dfreeman and/or @jamescdavis will be able to as well!

**Comment by RyanCavanaugh** [comment]
What's the status on this PR? I'd like to merge or close by the end of the week if possible

**Comment by chriskrycho** [comment]
Got buried, will review by Friday. Thanks for flagging it up!

**Comment by chriskrycho** [comment]
Sorry, ended up buried in one of *those* bugs where a build would fail only in a production build on a very specific circumstance that bedeviled us all day and concluded with us realizing that build configuration is irrelevant because it never gets executed and will in fact be impossible in another few weeks. It was that kind of day. :weary:

This is on my list for Monday morning, however!

**Comment by RyanCavanaugh** [comment]
lol no rush. I'll let next week's maintainer know

**Comment by ro0gr** [comment]
@chriskrycho @dfreeman @jamescdavis did you have a chance to take a look here?

**Comment by chriskrycho** [comment]
Doh! I got sucked into other things last week; I'm very sorry. It's on today's to-do list now!

**Comment by ro0gr** [comment]
No problem, I'll just keep pinging everyone until it's settled somehow! 😃

**Comment by chriskrycho** [comment]
Hahaha, PLEASE DO.

**Comment by elibarzilay** [comment]
@ro0gr: reading the above, I hate to do this, but it sounds like it should still wait for a proper review?

**Comment by ro0gr** [comment]
@elibarzilay yes, still waiting for review.

**Comment by jamescdavis** [comment]
Note: I'm actively reviewing this and testing it against a couple of large codebases. Thank you for your patience!

**Comment by weswigham** [comment]
@chriskrycho uhh... reminder ping, I guess

**Comment by chriskrycho** [comment]
@weswigham yes, thank you (sincerely)! @jamescdavis picked it up for me, and will hopefully have it knocked out this week.

**Comment by ro0gr** [comment]
I still believe this is a good step forward for ED types. However, taking into account, it still doesn't have any feedback, I assume, there are some issues with the PR, which I'm not aware of. I'm going to close it in a week if nothing changes. Otherwise, please, let me know if I can do smth else here. Thanks.

**Comment by chriskrycho** [comment]
So sorry about the delays. @jamescdavis do you still have time to dig into this?

**Comment by jamescdavis** [comment]
I worked on this today. I'm going to suggest some improvements that will give us better DX.

**Comment by jamescdavis** [comment]
There also seems to be problem with `model.belongsTo()`. I'll have to look a little closer in the morning.

**Review by sandersn — CHANGES_REQUESTED** [review]
Requesting changes so the bot understands the status.

**Comment by ro0gr** [comment]
@jamescdavis I'm curious about DX improvements and issues with the `model.belongsTo(` you've mentioned. Did you have a chance to take a look here?

**Comment by jamescdavis** [comment]
@ro0gr the DX improvement I'm thinking of would be to prevent the usage of model keys of the wrong type in `Model.belongsTo()` and `Model.hasMany()` rather than just narrowing the return type to `null`. For example, in [types/ember-data/test/belongs-to.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42839/files#diff-a8c90eb2deb65dbe6c333fd155f25444) you can currently say:
```ts
folder.belongsTo('name').value()!;
folder.belongsTo('children').value()!;
```
and neither of them are type errors, they just return `null`. IMHO, this is not as good at making specifying invalid keys a type error. One way to do this is to define:
```ts
type BelongsToKeys<Model extends DS.Model> = {
    [K in keyof Model]: UnboxBelongsTo<Model[K]> extends never ? never : K
}[keyof Model];
```
and use it as:
```ts
class Model extends Ember.Object {
    belongsTo<T extends Model, K extends BelongsToKeys<T>>(this: T, name: K): BelongsToReference<UnboxBelongsTo<T[K]>>;
}
```
The same can be done for `hasMany`.

**Comment by jamescdavis** [comment]
I'm still looking into the `belongsTo` issues. In general, I'm seeing some weirdness when using the types in an app with multiple models that reference each other, but I might just have something wrong. I'm going to poke at it a bit more today and let you know what I find.

**Comment by jamescdavis** [comment]
Another thing that would be useful it to export `BelongsToKeys`, `HasManyKeys`, and `KeysForRelationships`. I've [needed these before](https://github.com/CenterForOpenScience/ember-osf-web/blob/5bc18b27e241dd92257d257dd4867ec1234a35e5/lib/osf-components/addon/components/validated-input/checkboxes/component.ts#L12).

**Comment by ro0gr** [comment]
@jamescdavis well, I did few more attempts to enable more strict mode with `BelongsToKeys`, but still have no luck. I'm not sure whether it's because of my TS-fu, or some TS issues, but there are diffferent issues appearing in different contexts. Arguably I can try keep fixes it there and there, but `this.belongsTo(` and `this.hasMany(` becomes broken again with this change, and it doesn't seem it's feasible to fix from our side.

I do still believe the current `unknown` solution is pretty good to go. At the end, it still disallows you to misuse references, while fixing model types inference and `this` with references.

**Comment by ro0gr** [comment]
@jamescdavis I just don't want to give @typescript-bot a pleasure to kill the PR, I'd better do it myself if needed 💥 So let me double check. 

Do you still think `BelongsToKeys`, `HasManyKeys`, and `KeysForRelationships` is the must here?

**Comment by elibarzilay** [comment]
I'm disappointed on behalf of @typescript-bot.

**Comment by jamescdavis** [comment]
@ro0gr I definitely want to take another look at this. I was unfortunately not able to give it the attention it deserved before it timed out. I'm on family vacation this week, but should be able take some time next week to see if we can't get it there.

**Pull request closed without merging**